src/tactic/remove_subs.lean

```diff
@@ -0,0 +1,160 @@
+/-
+Copyright (c) 2022 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+import tactic.linarith
+
+/-!
+# `remove_subs` -- a tactic for splitting `ℕ`-subtractions
+
+The tactic `remove_subs` looks for `ℕ`-subtractions in the goal and recursively replaces any
+subexpression of the form `a - b` by two branches: one where `a ≤ b` and `a - b` is replaced
+by `0` and one where `∃ c : ℕ, a = b + c` and it tries to replace `a` by `b + c`.
+
+If called with the optional `!`-flag, then `remove_subs!` tries `linarith` on all remaining goals.
+
+See the tactic-doc for further details.
+
+##  Todo
+*  extend the scope of `remove_subs` to include also `max/min`?  possibly even `split_ifs`?
+-/
+
+namespace tactic
+namespace remove_subs
+
+lemma nat.le_cases (a b : ℕ) : a - b = 0 ∨ ∃ c, a = b + c :=
+begin
+  by_cases ab : a ≤ b,
+  { exact or.inl (tsub_eq_zero_of_le ab) },
+  { rcases nat.exists_eq_add_of_le (not_le.mp ab).le with ⟨c, rfl⟩,
+    exact or.inr ⟨_, rfl⟩ },
+end
+
+/--  A convenience definition: `rw_at` rewrites either at a hypothesis or at the target. -/
+meta def rw_at : option name → expr → tactic unit
+| none      lem := rewrite_target lem
+| (some na) lem := get_local na >>= rewrite_hyp lem >> skip
+
+/--  Given a list `l` of pairs of expressions, `local_constants_last l` reorders the list `l`
+so that all pairs `(a,b) ∈ l` with `a` a local constant appear last.  This is used in `get_sub`
+so that the replacement of the `ℕ`-subtractions begins with subtractions where an old term
+can be substituted, rather than simply rewritten. -/
+meta def local_constants_last (l : list (expr × expr)) : list (expr × expr) :=
+let (csts, not_csts) := l.partition (λ e : expr × expr, e.1.is_local_constant) in not_csts ++ csts
+
+/--  `get_sub lo e` extracts a list of pairs `(a, b)` from the expression `e`, where `a - b` is a
+subexpression of `e`.  It also takes argument `lo`, informing it that
+* the initial input `e` is hypothesis `lo` and rewrites it to reduce the number of `ℕ`-subtractions,
+  using the identity `a - b - c = a - (b + c)`;
+* the subtractions that get extracted are subtractions in `ℕ`.
+-/
+meta def get_sub (lo : option name) : expr → tactic (list (expr × expr))
+| `(%%a - %%b - %%c) := do bc ← to_expr ``(%%b + %%c),
+                           to_expr ``(nat.sub_sub %%a %%b %%c) >>= rw_at lo,
+                           [ga, gb, gc] ← [a, b, c].mmap get_sub,
+                           return $ local_constants_last ((a, bc) :: (ga ++ gb ++ gc))
```

Comment:
Why do you need `local_constants_last` to be called at every step of the recursion here? Can't the caller do it?

Reply:
Very good point!  Initially, I had in mind of doing more pre-processing on the list.  However, the tactic appears fast and the bottleneck seems to be the use of `linarith`.  I moved the reordering to later on, but there is actually no test that fails without it.